src/datachain/query/dataset.py

```diff
@@ -292,23 +292,21 @@ def q(*columns):
 
 @frozen
 class Subtract(DatasetDiffOperation):
-    """
-    Calculates rows that are in a source query but are not in target query (diff)
-    This can be used to do delta updates (calculate UDF only on newly added rows)
-    Example:
-        >>> ds = DatasetQuery(name="dogs_cats") # some older dataset with embeddings
-        >>> ds_updated = (
-                DatasetQuery("gs://dvcx-datalakes/dogs-and-cats")
-                .filter(C.size > 1000) # we can also filter out source query
-                .subtract(ds)
-                .add_signals(calc_embeddings) # calculae embeddings only on new rows
-                .union(ds) # union with old dataset that's missing new rows
-                .save("dogs_cats_updated")
-            )
-    """
+    on: Sequence[str]
 
     def query(self, source_query: Select, target_query: Select) -> Select:
-        return self.catalog.warehouse.subtract_query(source_query, target_query)
+        sq = source_query.alias("source_query")
+        tq = target_query.alias("target_query")
+        on_clause = sqlalchemy.and_(
+            getattr(sq.c, col_name) == getattr(tq.c, col_name) for col_name in self.on
+        )  # type: ignore[arg-type]
+        outer_join = sqlalchemy.join(sq, tq, on_clause, isouter=True)
+
+        where_clause = sqlalchemy.and_(
+            getattr(tq.c, col_name) == None  # noqa: E711
```

Comment:
Because of this test fails in SaaS as in Clickhouse we don't have `null`-able columns so value of empty column is empty value of a specific type (for `str` is `""`).  We do have a static method to get default values for each sql type, e.g:
```python
from datachain.sql.types import String
String.default_value()  # this will be None in CLI and "" in SaaS
```
I'm not sure if it's easy to get these SQL types from signals schema though ... if we have python types only then we prob have a map somwehere between python types and SQL types that can be used.